packages/react/src/components/Popover/index.js

```diff
@@ -23,20 +32,142 @@ const Popover = React.forwardRef(function Popover(props, ref) {
 
+  useIsomorphicEffect(() => {
+    if (!autoAlign) {
+      setAutoAligned(false);
+      return;
+    }
+
+    if (!floating.current) {
+      return;
+    }
+
+    if (autoAligned === true) {
+      return;
+    }
+
+    const rect = floating.current.getBoundingClientRect();
+
+    // The conditions, per side, of when the popover is not visible, excluding the popover's internal padding(16)
+    const conditions = {
+      left: rect.x < -16,
+      top: rect.y < -16,
+      right: rect.x + (rect.width - 16) > document.documentElement.clientWidth,
+      bottom:
+        rect.y + (rect.height - 16) > document.documentElement.clientHeight,
+    };
+
+    if (
+      !conditions.left &&
+      !conditions.top &&
+      !conditions.right &&
+      !conditions.bottom
+    ) {
+      setAutoAligned(false);
+      return;
+    }
+
+    const alignments = [
+      'top',
+      'top-left',
+      'right-bottom',
+      'right',
+      'right-top',
+      'bottom-left',
+      'bottom',
+      'bottom-right',
+      'left-top',
+      'left',
+      'left-bottom',
+      'top-right',
+    ];
+
+    // Creates the prioritized list of options depending on ideal alignment coming from `align`
+    const options = [align];
+    let option =
+      alignments[(alignments.indexOf(align) + 1) % alignments.length];
+
+    while (option) {
+      if (options.includes(option)) {
+        break;
+      }
+      options.push(option);
+      option = alignments[(alignments.indexOf(option) + 1) % alignments.length];
+    }
+
+    function isVisible(alignment) {
+      popover.current.classList.add(`${prefix}--popover--${alignment}`);
+
+      const rect = floating.current.getBoundingClientRect();
+
+      // Check if popover is not visible to the left of the screen
+      if (rect.x < -16) {
+        popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+        return false;
+      }
+
+      // Check if popover is not visible at the top of the screen
+      if (rect.y < -16) {
+        popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+        return false;
+      }
+
+      // Check if popover is not visible to right of screen
+      if (rect.x + (rect.width - 16) > document.documentElement.clientWidth) {
+        popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+        return false;
+      }
+
+      // Check if popover is not visible to bottom of screen
+      if (rect.y + (rect.height - 16) > document.documentElement.clientHeight) {
+        popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+        return false;
+      }
+
+      popover.current.classList.remove(`${prefix}--popover--${alignment}`);
+      return true;
+    }
+
+    let alignment = null;
+
+    for (let i = 0; i < options.length; i++) {
+      const option = options[i];
+
+      if (isVisible(option)) {
+        alignment = option;
+        break;
+      }
+      alignment = align;
+    }
+
+    if (alignment) {
```

Comment:
Will `alignment` always be defined due to L156 above? Or is there a case where it's still `null`?